analyzers/src/SonarAnalyzer.CSharp/Rules/InfiniteRecursion_RoslynCfg.cs

```diff
@@ -0,0 +1,224 @@
+                        if (!walker.CheckPaths() && !cfg.ExitBlock.IsReachable)
+                        {
+                            c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, accessor.Keyword.GetLocation(), "property accessor's recursion"));
+                        }
+                    }
+                }
+            }
+
+            public void CheckForNoExitMethod(SyntaxNodeAnalysisContext c, CSharpSyntaxNode body, SyntaxToken identifier, ISymbol symbol)
+            {
+                var operation = c.SemanticModel.GetOperation(body.Parent);
+                var wrapper = new IOperationWrapperSonar(operation);
+                while (wrapper.Parent != null)
+                {
+                    wrapper = new IOperationWrapperSonar(wrapper.Parent);
+                }
+
+                var cfg = ControlFlowGraph.Create(wrapper.Instance.Syntax, c.SemanticModel);
+                var enclosingKinds = new HashSet<SyntaxKind>
+                {
+                    SyntaxKindEx.LocalFunctionStatement, SyntaxKind.SimpleLambdaExpression, SyntaxKind.AnonymousMethodExpression, SyntaxKind.ParenthesizedLambdaExpression
+                };
+                if (c.Node.IsKind(SyntaxKindEx.LocalFunctionStatement))
+                {
+                    // we need to go up and track all possible enclosing local function statements
+                    foreach (var enclosingFunction in c.Node.Ancestors().Where(x => enclosingKinds.Contains(x.Kind())).Reverse())
+                    {
+                        if (enclosingFunction.IsAnyKind(SyntaxKindEx.LocalFunctionStatement))
+                        {
+                            cfg = cfg.GetLocalFunctionControlFlowGraph(c.SemanticModel.GetDeclaredSymbol(enclosingFunction) as IMethodSymbol);
+                        }
+                        else
+                        {
+                            var operationWrapper = cfg.FlowAnonymousFunctionOperations().Single(x => x.WrappedOperation.Syntax == enclosingFunction);
+                            cfg = cfg.GetAnonymousFunctionControlFlowGraph(operationWrapper);
+                        }
+                    }
+
+                    cfg = cfg.GetLocalFunctionControlFlowGraph(symbol as IMethodSymbol);
+                }
+
+                var walker = new CfgWalkerForMethod(new RecursionAnalysisContext<ControlFlowGraph>(cfg, symbol, identifier.GetLocation(), c));
+                if (!walker.CheckPaths() && !cfg.ExitBlock.IsReachable)
+                {
+                    c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, identifier.GetLocation(), "method's recursion"));
+                }
+            }
+
+            private class CfgWalkerForMethod : CfgRecursionSearcher
+            {
+                private readonly BooleanBlockMatcher existReferenceToDeclaringSymbolMatcher;
+
+                public CfgWalkerForMethod(RecursionAnalysisContext<ControlFlowGraph> context)
```

Comment:
These two walkers are very syntax-based. But what they do as a core functionality is `BlockHasReferenceToDeclaringSymbol`. That one should be easy to rewrite using operation-based approach that could simplify a lot of stuff around here.